### PR TITLE
Implementation/42874 storage authorization frontend

### PR DESCRIPTION
**Pull request by apfohl**

### CODE REVIEW

File: frontend/src/app/shared/components/file-links/file-link-list/file-link-list.component.ts [hunk 1]
```diff
@@ -38,6 +38,9 @@ import { CurrentUserService } from 'core-app/core/current-user/current-user.serv
 import { UntilDestroyedMixin } from 'core-app/shared/helpers/angular/until-destroyed.mixin';
 import { I18nService } from 'core-app/core/i18n/i18n.service';
 import isNewResource from 'core-app/features/hal/helpers/is-new-resource';
+import { CookieService } from 'ngx-cookie-service';
+import { v4 as uuidv4 } from 'uuid';
```
Comment (Kharonus) [inline frontend/src/app/shared/components/file-links/file-link-list/file-link-list.component.ts]: minor: sorting of not `core-app` imports to top?

File: frontend/src/app/shared/components/file-links/file-link-list/file-link-list.component.ts [hunk 2]
```diff
@@ -158,33 +163,80 @@ export class FileLinkListComponent extends UntilDestroyedMixin implements OnInit
...
+    );
+
+    this.buttons.next([
+      new StorageActionButton(
+        this.text.infoBox.loginButton,
+        () => {
+          const nonce = uuidv4();
+          this.setAuthorizationCallbackCookie(nonce);
+          window.location.href = FileLinkListComponent.authorizationFailureActionUrl(
+            this.storage._links.authorize?.href as string,
```
Comment (Kharonus) [inline frontend/src/app/shared/components/file-links/file-link-list/file-link-list.component.ts]: can we guard this accessor? if authorize is not available? we could add an early return before creating the nonce.

File: frontend/src/app/shared/components/file-links/file-link-list/file-link-list.component.ts [hunk 3]
```diff
@@ -158,33 +163,80 @@ export class FileLinkListComponent extends UntilDestroyedMixin implements OnInit
...
+      this.text.infoBox.emptyStorageHeader,
+      this.text.infoBox.emptyStorageContent,
+      'add-link',
+    );
+
+    this.buttons.next([
+      new StorageActionButton(
+        this.text.infoBox.emptyStorageButton,
+        () => {
+          window.location.href = this.storageLocation;
```
Comment (Kharonus) [inline frontend/src/app/shared/components/file-links/file-link-list/file-link-list.component.ts]: use window.open() here, too, as we want to open the storage.

**Comment by apfohl** [comment]
We move the full storage authorization cycle spec to a new work package, as it will be needed to test different other things regarding the whole feature.

https://community.openproject.org/projects/openproject/work_packages/43001/activity

File: frontend/package.json [hunk 4]
```diff
@@ -140,6 +142,7 @@
     "ts-action-operators": "^9.1.2",
     "typedjson": "^1.5.1",
     "urijs": "^1.19.11",
+    "uuid": "^8.3.2",
```
Comment (HDinger) [inline frontend/package.json:145]: Why are there two different packages for UUID needed?

File: frontend/package.json [hunk 5]
```diff
@@ -128,6 +129,7 @@
     "ng-dynamic-component": "^6.0.0",
     "ng2-charts": "^2.3.1",
     "ng2-dragula": "^2.1.1",
+    "ngx-cookie-service": "^14.0.0",
```
Comment (HDinger) [inline frontend/package.json:132]: As I said before, when we use cookies we need to handle a cookie warning for the user. If that is out of scope for this ticket, please open a new one, so that we don't forget about it.

File: frontend/src/app/shared/components/file-links/file-link-list/file-link-list.component.ts [hunk 6]
```diff
@@ -138,8 +147,8 @@ export class FileLinkListComponent extends UntilDestroyedMixin implements OnInit
 
   private deriveStorageInformation(fileLinkCount:number):void {
     switch (this.storage._links.authorizationState.href) {
-      case 'urn:openproject-org:api:v3:storages:authorization:FailedAuthentication':
-        this.setAuthenticationFailureState(fileLinkCount);
+      case 'urn:openproject-org:api:v3:storages:authorization:FailedAuthorization':
```
Comment (HDinger) [inline frontend/src/app/shared/components/file-links/file-link-list/file-link-list.component.ts]: Once https://github.com/opf/openproject/pull/10859 is merged, these strings will be constants to avoid doubled maintenance. Please add / re-use them here once it is through.

File: frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html [hunk 7]
```diff
@@ -4,11 +4,11 @@
   <op-storage-information
     *ngIf="showInformationBox$ | async"
     class="op-files-tab--storage-info-box"
+    data-qa-selector="op-files-tab--storage-information"
     [infoIcon]="informationBoxIcon"
     [infoTextHeader]="informationBoxHeader"
     [infoTextContent]="informationBoxContent"
-    [buttonText]="informationBoxButton"
-    (buttonClickEventEmitter)="openStorageLocation()"
+    [buttons]="buttons | async"
```
Comment (HDinger) [inline frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html:11]: It feels a little bit weird to pass components via input to a child template. Why isn't the whole button logic in the `op-storage-information` component itself? From a gut feeling I'd say that it belongs to the component itself and not to the file-link-list. Further, with the current approach you can basically not re-use the component, as you'd need to duplicate the button logic.

File: frontend/src/app/shared/components/file-links/storage-information/storage-information.html [hunk 8]
```diff
@@ -18,10 +18,19 @@
 <div
   class="button-box"
 >
-  <button class="button button_no-margin">
-    <op-icon icon-classes="button--icon spot-icon spot-icon_external-link" class="op-icon--wrapper">
-      <i class="button--icon spot-icon spot-icon_external-link"></i>
+  <button *ngFor="let button of buttons"
```
Comment (HDinger) [inline frontend/src/app/shared/components/file-links/storage-information/storage-information.html:21]: Since I could not look at it in the browser, this is just guesswork, but I assume that two buttons next to each other won't have any space in between them.

**Review by HDinger — COMMENTED** [review]
I only have some minor remarks to the code 👍

File: frontend/package.json (hunk 5 above)
Comment (apfohl) [reply frontend/package.json:132]: I discussed this with @wielinde and @marcalcobe and we determined that this kind of cookie is a technical one and does not need to be further handled by a cookie warning.

File: frontend/package.json (hunk 4 above)
Comment (apfohl) [reply frontend/package.json:145]: What is the other one? This one is the actual dependency which was just added here because we now use it explicitly. It was already in the dependency tree prior to these changes. The other one is just typings and only a dev dependency.

File: frontend/src/app/shared/components/file-links/file-link-list/file-link-list.component.ts (hunk 6 above)
Comment (apfohl) [reply frontend/src/app/shared/components/file-links/file-link-list/file-link-list.component.ts]: I will if it is merge before this PR.

File: frontend/package.json (hunk 5 above)
Comment (HDinger) [reply frontend/package.json:132]: 👍

File: frontend/package.json (hunk 4 above)
Comment (HDinger) [reply frontend/package.json:145]: Ah okay, my bad. I just saw two additions in this file with `uuid` and assumed that they were two different packages. You can ignore this 👍

File: frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html (hunk 7 above)
Comment (apfohl) [reply frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html:11]: This is exactly the point why I introduced this model. I decoupled the logic of any button from it's implementation. I assumed for the start of this refactoring that a button has a label and some sort of action that is triggered once you click it. Therefore I created a model class that take exactly these two information. A string for the label and a function for the button action. The logic these buttons now have is passed in from the component that actually defines the use cases. In this case the `FileLinkListComponent`. This component just has a `StorageInformation` box. This component can now be completely dump and can be reused anywhere. The `FileLinkListComponent` defines in which state the label is and what the button does. Further it reduces code and complexity in the templates an keep it in the specific component.

Please contact me if I couldn't explain the concept well enough :).

Comment (HDinger) [reply frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html:11]: mhh I see. Then I have two other questions: 
1. Is the `storageInformationBox` useful without these buttons? (I assume yes, but want to double check)
2. Is it imaginable that the `storageInformationBox` will be used at another place with different buttons (or no buttons)? Or will it always be these buttons when the box is used?

File: frontend/src/app/shared/components/file-links/storage-information/storage-information.html (hunk 8 above)
Comment (apfohl) [reply frontend/src/app/shared/components/file-links/storage-information/storage-information.html:21]: Sadly there is no easy way in our Angular application to denote that there is just one button or nothing at all. I tried using a Promise for that case, but as I need to access multiple properties of the `StorageActionButton`, I would had to introduce the subscription logic in the component itself instead of using the `async` pipe for this.

But I agree, that if there were multiple buttons, they would look not correct. But we have no style for this. So i left it out until it was necessary. If someone puts two buttons there, then there will be a style how to do this.

File: frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html (hunk 7 above)
Comment (apfohl) [reply frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html:11]: 1. Yes we have a case where there is no button, when a connection error occurred to Nextcloud. Then it makes no sense to either open Nextcloud or retry authentication.
2. Yes, as this element was introduced with this feature but reside inside the tabs section, it will get reused at other places. And don"t have to be changed at all. Just define the array with the button models and done.

Further additions to the button model could be a function that can determine it the action can be executed (`canExecute:() => boolean`) and be bind to a disabled/enabled state of the button in the template. This is general a good concept to decouple logic from it's representation regarding interactable UI elements.

File: frontend/src/app/shared/components/file-links/storage-information/storage-information.html (hunk 8 above)
Comment (HDinger) [reply frontend/src/app/shared/components/file-links/storage-information/storage-information.html:21]: You don't need Angular for that. You can let the `button-box` handle that case via CSS. Either you display the box as a `grid` with `grid-column-gap: XYpx` or you add a `margin-right` to all buttons except for the last one (`:last-of-type`). https://github.com/opf/openproject/pull/10859 already introduces a class for the buttons and add some margins any way. So you could simply extend that (once it is merged)

File: frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html (hunk 7 above)
Comment (HDinger) [reply frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html:11]: okay, it's fine for me then 👍 Thanks for the explanation :)

File: frontend/src/app/shared/components/file-links/storage-information/storage-information.html (hunk 8 above)
Comment (apfohl) [reply frontend/src/app/shared/components/file-links/storage-information/storage-information.html:21]: Yes, that is true. I will do this.

File: frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html (hunk 7 above)
Comment (apfohl) [reply frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html:11]: Very welcome.

**Review by HDinger — COMMENTED** [review]
The code is fine for me. There are however two optimisations I'd like to see. 

1. Using the constants for the `urn:openproject-org:api...` strings.
2. Handlind the (styling) case that two buttons are shown next to each other.

However, both somehow interfere with https://github.com/opf/openproject/pull/10859 so you may want to wait for that one to avoid too many conflicts

**Review by Kharonus — APPROVED** [review]

**Review by HDinger — APPROVED** [review]
Lgtm 👍
